File: app/src/main/java/me/blog/korn123/easydiary/calendar/CalendarActivity.java
package me.blog.korn123.easydiary.calendar;

import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.roomorama.caldroid.CaldroidFragment;
import com.roomorama.caldroid.CaldroidListener;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.blog.korn123.commons.utils.DateUtils;
import me.blog.korn123.commons.utils.FontUtils;
import me.blog.korn123.easydiary.R;
import me.blog.korn123.easydiary.diary.DiaryDao;
import me.blog.korn123.easydiary.diary.DiaryDto;
import me.blog.korn123.easydiary.diary.DiarySimpleCardArrayAdapter;
import me.blog.korn123.easydiary.diary.ReadDiaryDetailActivity;
import me.blog.korn123.easydiary.helper.EasyDiaryActivity;

/**
 * Created by hanjoong on 2017-03-28.
 */

public class CalendarActivity extends EasyDiaryActivity {
    private CaldroidFragment caldroidFragment;

    @BindView(R.id.selectedList)
    ListView mSelectedListView;

    @BindView(R.id.emptyInfo)
    TextView mEmptyInfo;

    private ArrayAdapter<DiaryDto> mArrayAdapterDiary;
    private List<DiaryDto> mDiaryList;
    private Date mCurrentDate;

    private Date getCurrentDate() {
        if (mCurrentDate == null) mCurrentDate = Calendar.getInstance().getTime();
        return mCurrentDate;
    }

    @Override
    protected void onResume() {
        super.onResume();
        refreshList(getCurrentDate());
        caldroidFragment.refreshView();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_calendar);
        ButterKnife.bind(this);

        Toolbar toolbar = (Toolbar)findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(getString(R.string.calendar_title));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        FontUtils.setToolbarTypeface(toolbar, Typeface.DEFAULT);
        FontUtils.setTypefaceDefault(mEmptyInfo);

        Calendar cal = Calendar.getInstance();
        Date currentDate = cal.getTime();
        refreshList(currentDate);
        mArrayAdapterDiary = new DiarySimpleCardArrayAdapter(this, R.layout.list_item_diary_simple_card_array_adapter , this.mDiaryList);
        mSelectedListView.setAdapter(mArrayAdapterDiary);
        mSelectedListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                DiaryDto diaryDto = (DiaryDto)adapterView.getAdapter().getItem(i);
                Intent detailIntent = new Intent(CalendarActivity.this, ReadDiaryDetailActivity.class);
                detailIntent.putExtra("sequence", diaryDto.getSequence());
                detailIntent.putExtra("title", diaryDto.getTitle());
                detailIntent.putExtra("contents", diaryDto.getContents());
                detailIntent.putExtra("date", DateUtils.timeMillisToDateTime(diaryDto.getCurrentTimeMillis()));
                detailIntent.putExtra("current_time_millis", diaryDto.getCurrentTimeMillis());
                detailIntent.putExtra("weather", diaryDto.getWeather());
                startActivity(detailIntent);
            }
        });

        // Setup caldroid fragment
        // **** If you want normal CaldroidFragment, use below line ****
//        caldroidFragment = new CaldroidFragment();

        // //////////////////////////////////////////////////////////////////////
        // **** This is to show customized fragment. If you want customized
        // version, uncomment below line ****
        caldroidFragment = new CaldroidCustomFragment();

        // Setup arguments

        // If Activity is created after rotation
        if (savedInstanceState != null) {
            caldroidFragment.restoreStatesFromKey(savedInstanceState,
                    "CALDROID_SAVED_STATE");
        }
        // If activity is created from fresh
        else {
            Bundle args = new Bundle();
            args.putInt(CaldroidFragment.MONTH, cal.get(Calendar.MONTH) + 1);
            args.putInt(CaldroidFragment.YEAR, cal.get(Calendar.YEAR));
            args.putBoolean(CaldroidFragment.ENABLE_SWIPE, true);
            args.putBoolean(CaldroidFragment.SIX_WEEKS_IN_CALENDAR, true);

            // Uncomment this to customize startDayOfWeek
            // args.putInt(CaldroidFragment.START_DAY_OF_WEEK,
            // CaldroidFragment.TUESDAY); // Tuesday

            // Uncomment this line to use Caldroid in compact mode
            // args.putBoolean(CaldroidFragment.SQUARE_TEXT_VIEW_CELL, false);

            // Uncomment this line to use dark theme
//            args.putInt(CaldroidFragment.THEME_RESOURCE, com.caldroid.R.style.CaldroidDefaultDark);

            caldroidFragment.setArguments(args);
        }

        caldroidFragment.setSelectedDate(currentDate);

//        setCustomResourceForDates();

        // Attach to the activity
        FragmentTransaction t = getSupportFragmentManager().beginTransaction();
        t.replace(R.id.calendar1, caldroidFragment);
        t.commit();

        // Setup listener
        final CaldroidListener listener = new CaldroidListener() {

            @Override
            public void onSelectDate(Date date, View view) {
//                Toast.makeText(getApplicationContext(), formatter.format(date),
//                        Toast.LENGTH_SHORT).show();
//                ColorDrawable green = new ColorDrawable(Color.GREEN);
//                caldroidFragment.setBackgroundDrawableForDate(green, date);
                caldroidFragment.clearSelectedDates();
                caldroidFragment.setSelectedDate(date);
                caldroidFragment.refreshView();
                mCurrentDate = date;
                refreshList(date);
            }

            @Override
            public void onChangeMonth(int month, int year) {
                String text = "month: " + month + " year: " + year;
//                Toast.makeText(getApplicationContext(), text,
//                        Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onLongClickDate(Date date, View view) {
//                Toast.makeText(getApplicationContext(),
//                        "Long click " + formatter.format(date),
//                        Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onCaldroidViewCreated() {
                if (caldroidFragment.getLeftArrowButton() != null) {
//                    Toast.makeText(getApplicationContext(),
//                            "Caldroid view is created", Toast.LENGTH_SHORT)
//                            .show();
                }
            }

        };

        // Setup Caldroid
        caldroidFragment.setCaldroidListener(listener);
    }

    /**
     * Save current states of the Caldroid here
     */
    @Override
    protected void onSaveInstanceState(Bundle outState) {
        // TODO Auto-generated method stub
        super.onSaveInstanceState(outState);

        if (caldroidFragment != null) {
            caldroidFragment.saveStatesToKey(outState, "CALDROID_SAVED_STATE");
        }
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    public void refreshList(Date date) {
        final SimpleDateFormat formatter = new SimpleDateFormat(DateUtils.DATE_PATTERN_DASH);

        if (mDiaryList == null) {
            mDiaryList = DiaryDao.readDiaryByDateString(formatter.format(date));
        } else if (mDiaryList != null) {
            mDiaryList.clear();
            mDiaryList.addAll(DiaryDao.readDiaryByDateString(formatter.format(date)));
            mArrayAdapterDiary.notifyDataSetChanged();
        }

        if (mDiaryList.size() > 0) {
            mSelectedListView.setVisibility(View.VISIBLE);
            mEmptyInfo.setVisibility(View.GONE);
        } else {
            mSelectedListView.setVisibility(View.GONE);
            mEmptyInfo.setVisibility(View.VISIBLE);
        }
    }

    //    private void setCustomResourceForDates() {
//        Calendar cal = Calendar.getInstance();
//
//        Date currentDate = cal.getTime();
//
//        // Min date is last 7 days
//        cal.add(Calendar.DATE, -7);
//        Date blueDate = cal.getTime();
//
//        // Max date is next 7 days
//        cal = Calendar.getInstance();
//        cal.add(Calendar.DATE, 7);
//        Date greenDate = cal.getTime();
//
//        if (caldroidFragment != null) {
//            ColorDrawable blue = new ColorDrawable(getResources().getColor(R.color.blue));
//            ColorDrawable green = new ColorDrawable(Color.GREEN);
//            caldroidFragment.setBackgroundDrawableForDate(blue, blueDate);
//            caldroidFragment.setBackgroundDrawableForDate(green, greenDate);
//            caldroidFragment.setTextColorForDate(R.color.white, blueDate);
//            caldroidFragment.setTextColorForDate(R.color.white, greenDate);
//        }
//    }

}

File: app/src/main/java/me/blog/korn123/easydiary/activities/CalendarActivity.kt
package me.blog.korn123.easydiary.activities

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import com.roomorama.caldroid.CaldroidFragment
import com.roomorama.caldroid.CaldroidFragmentEx
import com.roomorama.caldroid.CaldroidListener
import io.realm.Sort
import me.blog.korn123.commons.utils.DateUtils
import me.blog.korn123.easydiary.R
import me.blog.korn123.easydiary.activities.DiaryWritingActivity.Companion.INITIALIZE_TIME_MILLIS
import me.blog.korn123.easydiary.adapters.DiaryCalendarItemAdapter
import me.blog.korn123.easydiary.databinding.ActivityCalendarBinding
import me.blog.korn123.easydiary.extensions.config
import me.blog.korn123.easydiary.extensions.dpToPixelFloatValue
import me.blog.korn123.easydiary.extensions.isLandScape
import me.blog.korn123.easydiary.fragments.CalendarFragment
import me.blog.korn123.easydiary.helper.CALENDAR_SORTING_ASC
import me.blog.korn123.easydiary.helper.DEFAULT_CALENDAR_FONT_SCALE
import me.blog.korn123.easydiary.helper.DIARY_SEQUENCE
import me.blog.korn123.easydiary.helper.EasyDiaryDbHelper
import me.blog.korn123.easydiary.helper.TransitionHelper
import me.blog.korn123.easydiary.models.Diary
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Formatter
import java.util.Locale

/**
 * Created by Hanjoong.Cho on 2017-03-28
 * Refactored code on 2020-01-03
 *
 */

class CalendarActivity : EasyDiaryActivity() {
    /***************************************************************************************************
     *   global properties
     *
     ***************************************************************************************************/
    private lateinit var mBinding: ActivityCalendarBinding
    private lateinit var mCalendarFragment: CaldroidFragmentEx
    private lateinit var mDatePickerDialog: DatePickerDialog
    private val mCalendar = Calendar.getInstance(Locale.getDefault())
    private var mDiaryList: MutableList<Diary> = mutableListOf()
    private var mArrayAdapterDiary: ArrayAdapter<Diary>? = null

    /***************************************************************************************************
     *   override functions
     *
     ***************************************************************************************************/
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = ActivityCalendarBinding.inflate(layoutInflater)
        setContentView(mBinding.root)
        if (config.settingCalendarFontScale != DEFAULT_CALENDAR_FONT_SCALE && !isLandScape()) {
            mBinding.calendarCard.layoutParams =
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 6F)
            mBinding.frameBottom?.layoutParams =
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 4F)
        }
        setSupportActionBar(mBinding.toolbar)
        supportActionBar?.run {
            title = getString(R.string.calendar_title)
            setDisplayHomeAsUpEnabled(true)
        }

        mDatePickerDialog =
            DatePickerDialog(
                this,
                { _, year, month, dayOfMonth ->
                    mCalendar.set(year, month, dayOfMonth)
                    mCalendarFragment.moveToDate(mCalendar.time)
                    selectDateAndRefreshView()
                },
                mCalendar.get(Calendar.YEAR),
                mCalendar.get(Calendar.MONTH),
                mCalendar.get(Calendar.DAY_OF_MONTH),
            )

        if (config.enableCardViewPolicy) {
            mBinding.calendarCard.useCompatPadding = true
            mBinding.calendarCard.cardElevation = dpToPixelFloatValue(2F)
        } else {
            mBinding.calendarCard.useCompatPadding = false
            mBinding.calendarCard.cardElevation = 0F
        }

        val cal = Calendar.getInstance()
        val currentDate = cal.time
        refreshList()
        mArrayAdapterDiary =
            DiaryCalendarItemAdapter(this, R.layout.item_diary_calendar, this.mDiaryList)
        mBinding.selectedList.adapter = mArrayAdapterDiary
        mBinding.selectedList.onItemClickListener =
            AdapterView.OnItemClickListener { adapterView, view, i, l ->
                val diaryDto = adapterView.adapter.getItem(i) as Diary
                val detailIntent = Intent(this@CalendarActivity, DiaryReadingActivity::class.java)
                detailIntent.putExtra(DIARY_SEQUENCE, diaryDto.sequence)
                TransitionHelper.startActivityWithTransition(this@CalendarActivity, detailIntent)
            }

        mCalendarFragment = CalendarFragment()

        // Setup arguments
        // If Activity is created after rotation
        if (savedInstanceState != null) {
            mCalendarFragment.restoreStatesFromKey(savedInstanceState, "CALDROID_SAVED_STATE")
        } else {
            val args = Bundle()
            args.putInt(CaldroidFragment.MONTH, cal.get(Calendar.MONTH) + 1)
            args.putInt(CaldroidFragment.YEAR, cal.get(Calendar.YEAR))
            args.putInt(CaldroidFragment.START_DAY_OF_WEEK, config.calendarStartDay)
            args.putBoolean(CaldroidFragment.ENABLE_SWIPE, true)
            args.putBoolean(CaldroidFragment.SIX_WEEKS_IN_CALENDAR, true)

            mCalendarFragment.arguments = args
        }

        mCalendarFragment.setSelectedDate(currentDate)

        // Attach to the activity
        val t = supportFragmentManager.beginTransaction()
        t.replace(R.id.calendar1, mCalendarFragment)
        t.commit()

        mCalendarFragment.caldroidListener =
            object : CaldroidListener() {
                override fun onSelectDate(
                    date: Date,
                    view: View,
                ) {
                    mCalendar.time = date
                    syncDatePicker()
                    selectDateAndRefreshView()
                }

                override fun onChangeMonth(
                    month: Int,
                    year: Int,
                ) {
                    val monthYearFlag =
                        android.text.format.DateUtils.FORMAT_SHOW_DATE or android.text.format.DateUtils.FORMAT_NO_MONTH_DAY or android.text.format.DateUtils.FORMAT_SHOW_YEAR
                    val monthYearFormatter = Formatter(StringBuilder(50), Locale.getDefault())
                    val calendar = Calendar.getInstance(Locale.getDefault())
                    calendar.set(Calendar.YEAR, year)
                    calendar.set(Calendar.MONTH, month - 1)
                    calendar.set(Calendar.DATE, 1)
                    val monthTitle =
                        android.text.format.DateUtils
                            .formatDateRange(
                                this@CalendarActivity,
                                monthYearFormatter,
                                calendar.timeInMillis,
                                calendar.timeInMillis,
                                monthYearFlag,
                            ).toString()
                    supportActionBar?.subtitle = monthTitle.uppercase(Locale.getDefault())
                }

                override fun onLongClickDate(
                    date: Date?,
                    view: View?,
                ) {}

                override fun onCaldroidViewCreated() {}
            }

        mBinding.writeDiary.setOnClickListener {
            TransitionHelper.startActivityWithTransition(
                this,
                Intent(this, DiaryWritingActivity::class.java).apply {
                    putExtra(INITIALIZE_TIME_MILLIS, mCalendar.timeInMillis)
                },
            )
        }
    }

    override fun onResume() {
        super.onResume()
        refreshList()
        mCalendarFragment.refreshView()
    }

    /**
     * Save current states of the Caldroid here
     */
    override fun onSaveInstanceState(outState: Bundle) {
        // TODO Auto-generated method stub
        super.onSaveInstanceState(outState)
        mCalendarFragment.saveStatesToKey(outState, "CALDROID_SAVED_STATE")
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_calendar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.previous -> mCalendarFragment.prevMonth()
            R.id.next -> mCalendarFragment.nextMonth()
            R.id.datePicker -> mDatePickerDialog.show()
        }
        return super.onOptionsItemSelected(item)
    }

    /***************************************************************************************************
     *   etc functions
     *
     ***************************************************************************************************/
    private fun refreshList() {
        val formatter = SimpleDateFormat(DateUtils.DATE_PATTERN_DASH, Locale.getDefault())
        val sort: Sort =
            if (config.calendarSorting == CALENDAR_SORTING_ASC) Sort.ASCENDING else Sort.DESCENDING
        mDiaryList.clear()
        mDiaryList.addAll(
            EasyDiaryDbHelper.findDiaryByDateString(
                formatter.format(mCalendar.time),
                sort,
            ),
        )
        mArrayAdapterDiary?.notifyDataSetChanged()
        mBinding.selectedList.setSelection(0)

        if (mDiaryList.size > 0) {
            mBinding.selectedList.visibility = View.VISIBLE
            mBinding.emptyInfo.visibility = View.GONE
        } else {
            mBinding.selectedList.visibility = View.GONE
            mBinding.emptyInfo.visibility = View.VISIBLE
        }
    }

    private fun selectDateAndRefreshView() {
        mCalendarFragment.clearSelectedDates()
        mCalendarFragment.setSelectedDate(mCalendar.time)
        mCalendarFragment.refreshViewOnlyCurrentPage()
        refreshList()
    }

    private fun syncDatePicker() {
        mDatePickerDialog.updateDate(
            mCalendar.get(Calendar.YEAR),
            mCalendar.get(Calendar.MONTH),
            mCalendar.get(Calendar.DAY_OF_MONTH),
        )
    }
}
